File: HomeWork/src/main/python/week05/solution1011.py
from typing import List


class Solution:
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        def check(weights: List[int], days: int, box: int):
            day = 0
            has_weight = 0
            for weight in weights:
                has_weight += weight
                if has_weight > box:
                    day += 1
                    if day > days:
                        return False
                    has_weight = weight

            return day < days

        left = max(weights)
        right = sum(weights)

        while left < right:
            mid = (left + right) // 2

            if check(weights, days, mid):
                right = mid
            else:
                left = mid + 1

        return right

File: HomeWork/src/main/java/week05/Solution1011.java
package week05;

public class Solution1011 {
    int[] weights;
    int days;

    public int shipWithinDays(int[] weights, int days) {
        this.weights = weights;
        this.days = days;

        int left = 0;
        int right = 0;

        for (int weight : weights) {
            right += weight;

            if (weight > left) {
                left = weight;
            }
        }

        while (left < right) {
            int mid = (left + right) / 2;

            if (check(mid)) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return right;
    }

    private boolean check(int box) {
        int useDay = 0;
        int hasWeight = 0;
        for (int weight : weights) {
            hasWeight += weight;
            if (hasWeight > box) {
                useDay += 1;
                if (useDay > days) {
                    return false;
                }
                hasWeight = weight;
            }
        }
        return useDay < days;

    }

    public static void main(String[] args) {
        Solution1011 s = new Solution1011();
        int[] weights = { 1, 2, 3, 1, 1 };
        int days = 4;

        int res = s.shipWithinDays(weights, days);
        System.out.println(res);
    }

}
